profiling/src/lib.rs

```diff
@@ -934,66 +1030,112 @@ extern "C" fn execute_internal(
+/// - `zend_mm_gc()` will not do anything anymore.
+/// - `zend_mm_shutdown()` wont cleanup chunks anymore, leading to memory leaks
+/// The `_zend_mm_heap`-struct itself is private, but we are lucky, as the `use_custom_heap` flag
+/// is the first element and thus the first 4 bytes.
+/// Take care and call `restore_zend_heap()` afterwards!
+#[cfg(feature = "allocation_profiling")]
+unsafe fn prepare_zend_heap(heap: *mut zend::_zend_mm_heap) -> c_int {
+    let custom_heap: c_int = std::ptr::read(heap as *const c_int);
+    std::ptr::write(heap as *mut c_int, zend::ZEND_MM_CUSTOM_HEAP_NONE as c_int);
+    custom_heap
+}
+
+/// Restore the ZendMM heap's `use_custom_heap` flag, see `prepare_zend_heap` for details
+#[cfg(feature = "allocation_profiling")]
+unsafe fn restore_zend_heap(heap: *mut zend::_zend_mm_heap, custom_heap: c_int) {
+    std::ptr::write(heap as *mut c_int, custom_heap);
+}
+
+#[cfg(feature = "allocation_profiling")]
 unsafe extern "C" fn alloc_profiling_malloc(len: u64) -> *mut ::libc::c_void {
     let ptr: *mut libc::c_void;
 
     // TODO: prepare a function pointer to use so we don't need a runtime check
     if PREV_CUSTOM_MM_ALLOC.is_none() {
-        ptr = zend::_zend_mm_alloc(zend::zend_mm_get_heap(), len);
+        let heap = zend::zend_mm_get_heap();
+        let custom_heap = prepare_zend_heap(heap);
+        ptr = zend::_zend_mm_alloc(heap, len);
+        restore_zend_heap(heap, custom_heap);
     } else {
         let prev = PREV_CUSTOM_MM_ALLOC.unwrap();
         ptr = prev(len);
     }
 
     // during startup, minit, rinit, ... current_execute_data is null
-    if zend::executor_globals.current_execute_data.is_null() {
+    // we are only interested in allocations during userland operations
+    if zend::ddog_php_prof_get_current_execute_data().is_null() {
         return ptr;
     }
 
-    REQUEST_LOCALS.with(|cell| {
-        // Panic: there might already be a mutable reference to `REQUEST_LOCALS`
-        let locals = cell.try_borrow();
-        if locals.is_err() {
-            return;
-        }
-        let locals = locals.unwrap();
-
-        if !locals.profiling_enabled {
-            return;
-        }
-
-        if let Some(profiler) = PROFILER.lock().unwrap().as_ref() {
-            // Safety: execute_data was provided by the engine, and the profiler doesn't mutate it.
-            unsafe {
-                profiler.collect_allocations(
-                    zend::executor_globals.current_execute_data,
-                    len,
-                    &locals,
-                )
-            };
-        }
+    ALLOCATION_PROFILING_STATS.with(|cell| {
+        let mut allocations = cell.borrow_mut();
+        allocations.track_allocation(len)
     });
 
     ptr
 }
 
+// This function exists just for completeness. When calling `zend_mm_set_custom_handlers()` you
+// need to pass a pointer to a `free()` function.
+#[cfg(feature = "allocation_profiling")]
```

Comment:
I didn't understand your comment here about completeness. Can you explain some more?

Reply:
I adjusted the comment in a new commit.
The thing is: `zend_mm_set_custom_handlers()` expects three function pointers. If any of these is `null` it would unregister custom handlers. So although we are - currently - not interested in tracking `free()` calls, we need to provide a function pointer.
As there will be a custom handler installed, we need to overwrite the `use_custom_heap` flag before calling the ZendMM default free function.
Hence the only reason this function exists is because we are using the `zend_mm_set_custom_handlers()` hook.